### PR TITLE
accounts-db: test_hash_stored_account: Avoid UB.

### PR DESCRIPTION
**Pull request by ilya-bobyr**
`unsafe { transmute }` in the test, as written, is undefined behavior. And, I think, we actually see the compiler generating different code for it as we are changing the generated hashes when we upgrade the compiler. It happened during upgrade to 1.67.1:

  https://github.com/solana-labs/solana/pull/29947

And the hashes changed again in 1.71.

---

Same as https://github.com/solana-labs/solana/pull/33082, but numbers are sequential.

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -12551,32 +12551,35 @@ pub mod tests {
...
-
-        // spray memory with decreasing counts so that, data layout can be detected.
-        for (i, byte) in blob.iter_mut().enumerate() {
-            *byte = (INPUT_LEN - i) as u8;
-        }
-
-        //UNSAFE: forcibly cast the special byte pattern to actual account fields.
-        let (slot, meta, account_meta, data, offset, hash): InputTuple =
-            unsafe { std::mem::transmute::<InputBlob, InputTuple>(blob) };
+        let data: [u8; ACCOUNT_DATA_LEN] = [0x69, 0x6b, 0x6c];
```
Comment (apfitzge) [inline accounts-db/src/accounts_db.rs:12576]: 0x6a?

**Comment by apfitzge** [comment]
Can you explain why there is UB in the first place?
Is the UB coming from the byte of our `executable` bool not being 0 or 1?

**Comment by ilya-bobyr** [comment]
> Can you explain why there is UB in the first place? Is the UB coming from the byte of our `executable` bool not being 0 or 1?

I thought that producing a value from bytes for types with Rust representation is undefined behavior.  Because it writes into padding.
While it is based on some valid assumptions, this is an incorrect conclusion.
It is only an undefined behavior to write into padding only through a pointer.
And `transmute` explicitly says that it may or may not populate the padding part - and that is OK.

So, it only leaves the undefined behavior of construction of an invalid `bool` value.

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (ilya-bobyr) [reply accounts-db/src/accounts_db.rs:12576]: Good catch :)

**Comment by apfitzge** [comment]
Is the UB on the bool what causes us to have different hashes for debug vs release?

I think I'm fine with this PR if we can remove that annoyance as well. In append_vec we sanitize `executable` to ensure higher 7-bits are 0 on loading, so testing that here is unnecessary. maybe we shouldn't have had a bool in the first place and just had a u8 bit-flags instead so we didn't even have to think about all this weirdness around bool.

**Comment by ilya-bobyr** [comment]
> Is the UB on the bool what causes us to have different hashes for debug vs release?

I think so.

After `repr(C)` was added, the layout should be consistent, otherwise it is no good for ABIs.
It is not something that can be optimized :)

At first, I thought, it might be that something deeper does not have a proper layout specifier.
Manual inspection showed no offenders.
Then I found there is a compiler attribute that makes the compiler print the layout information.

I've extracted all the relevant types, and you can run it in the playground now:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=aa9bcb4af338cb621195bbfbc0a33d88

It shows identical layout between debug and release.

Considering that when I remove `unsafe`, results are consistent between debug and release, I really do not have any explanations, except for the `bool`.

`hash_account_data()` actually uses the `executable` value:

```rust
        if executable {
            hasher.update(&[1u8; 1]);
        } else {
            hasher.update(&[0u8; 1]);
        }
```

This can be written differently by the debug and release compilation passes.
And as we pass in a value that is not 0 or 1, we are probably adding this value directly into the hash, rather than passing in 0 or 1.

---

Here goes a detailed explanation of the UB for `bool`.
Just to make sure we are on the same page :)

My understanding is that this UB is there so that the compiler could write assembly, always assuming that a bool is either 0 or 1.
I do not know the exact patters it uses.
But I can imagine a number of cases where a different value would cause all kinds of issues.

For example

```rust
pub fn f(three_not_two: bool, mut val: u64) -> u64 {
    if three_not_two {
        val += 3;
    } else {
        val += 2;
    }
    val
}
```
with `-O` compiles to
```asm
example::f:
        mov     eax, edi
        add     rax, rsi
        add     rax, 2
        ret
```
https://godbolt.org/z/4h3eerG11

If this rule is broken, combined with optimizations, it can lead to very strange results.
Another strange result can happen, if we use the resulting value as a jump offset.

> I think I'm fine with this PR if we can remove that annoyance as well. In append_vec we sanitize `executable` to ensure higher 7-bits are 0 on loading, so testing that here is unnecessary. maybe we shouldn't have had a bool in the first place and just had a u8 bit-flags instead so we didn't even have to think about all this weirdness around bool.

In my mind, this is exactly the reason not to use disable compiler checks with `unsafe`.
If there is a way to do it that requires less mental load, in most cases, it will save time for everyone involved.

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (ilya-bobyr) [reply accounts-db/src/accounts_db.rs]: Fixed.

**Comment by ilya-bobyr** [comment]
Rebased and added the lost `0x6a`.

**Review by apfitzge — DISMISSED** [review]
lgtm, previously relying on UB for the hashes because of the executable. Use manually specified bytes for non-bools and set the bool explicitly to false addresses this. No more need for different hashes between debug and release

**Comment by ryoqun** [comment]
> Then I found there is a compiler attribute that makes the compiler print the layout information.

oh, `#[rustc_layout(debug)]` in `#![feature(rustc_attrs)]` sounds interesting. :) maybe, this is what rust-analyzer is using internally.

> Here goes a detailed explanation of the UB for bool.
> ...

thanks for good write up. yeah, that aligns with my understanding.

**Review by ryoqun — DISMISSED** [review]
lgtm with nits; thanks for writing a proper test, rebasing and some write-ups to convince the team.

I appreciate your willingness for removing `unsafe`s or UBs. :)

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -12561,40 +12561,36 @@ pub mod tests {
 
     #[test]
     fn test_hash_stored_account() {
-        // This test uses some UNSAFE tricks to detect most of hashing code changes, resulting from
-        // account's field additions and deletions of StoredAccountMeta and AccountSharedData and
-        // hashing-order changes.
-
+        // Number are just sequential.
+        let slot: Slot = 0x0102030405060708;
```
Comment (ryoqun) [inline accounts-db/src/accounts_db.rs:12565]: nit: btw, you can place `_` arbitrarily: ie. `0x01_02_03_04_05_06_07_08`

File: accounts-db/src/accounts_db.rs [hunk 3]
```diff
@@ -12561,40 +12561,36 @@ pub mod tests {
...
-        type InputBlob = [u8; INPUT_LEN];
-        let mut blob: InputBlob = [0u8; INPUT_LEN];
-
-        // spray memory with decreasing integers so that, data layout change and/or hashing
-        // reordering can be detected. note that just zeroed blob can't detect field reordering.
-        for (i, byte) in blob.iter_mut().enumerate() {
-            *byte = (INPUT_LEN - i) as u8;
-        }
+        let data: [u8; ACCOUNT_DATA_LEN] = [0x69, 0x6a, 0x6b];
+        let offset: usize = 0x6c6d6e6970717273;
```
Comment (ryoqun) [inline accounts-db/src/accounts_db.rs:12587]: nit: `0x6c6d6e6**9**70717273` => `0x6c6d6e6**f**70717273`

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (ilya-bobyr) [reply accounts-db/src/accounts_db.rs:12565]: Added underscores.

File: accounts-db/src/accounts_db.rs (hunk 3 above)
Comment (ilya-bobyr) [reply accounts-db/src/accounts_db.rs:12587]: Fixed.

**Comment by ilya-bobyr** [comment]
Added underscores and corrected the missing `6f` byte.
Rebased on top of the latest `master`.

**Comment by ilya-bobyr** [comment]
> > Then I found there is a compiler attribute that makes the compiler print the layout information.
> 
> oh, `#[rustc_layout(debug)]` in `#![feature(rustc_attrs)]` sounds interesting. :) maybe, this is what rust-analyzer is using internally.

What `rust-analyzer` functionality are you referring to?

**Review by apfitzge — APPROVED** [review]
lgtm

**Comment by ryoqun** [comment]
> What `rust-analyzer` functionality are you referring to?

this one: https://rust-analyzer.github.io/thisweek/2023/07/10/changelog-189.html#new-features

and lightly touched in our learning-corner as well: https://discord.com/channels/428295358100013066/977244255212937306/1146863043578445967

**Pull request merged**